tensorflow/c/kernels/merge_summary_op.cc

```diff
@@ -0,0 +1,128 @@
+
+#include <sstream>
+#include <unordered_set>
+
+#include "tensorflow/c/kernels.h"
+#include "tensorflow/c/tf_tensor.h"
+#include "tensorflow/c/tf_status.h"
+#include "tensorflow/core/framework/selective_registration.h"
+#include "tensorflow/core/framework/summary.pb.h"
+#include "tensorflow/core/framework/types.h"
+#include "tensorflow/core/platform/macros.h"
+#include "tensorflow/core/platform/protobuf.h"
+#include "tensorflow/core/platform/tstring.h"
+#include "tensorflow/core/platform/default/logging.h"
+
+// Struct that stores the status and TF_Tensor inputs to the opkernel. 
+// Used to delete tensor and status in its destructor upon kernel return. 
+struct TensorWrapper { 
+  TF_Tensor* t; 
+  TensorWrapper() : t(nullptr) {}
+  ~TensorWrapper() { 
+    TF_DeleteTensor(t);
+  }
+};
+
+struct Status_Wrapper { 
+  TF_Status* s; 
+  Status_Wrapper() { 
+    s = TF_NewStatus(); 
+  }
+  ~Status_Wrapper() { 
+    TF_DeleteStatus(s);
+  }
+};
+
+// dummy functions used for kernel registration 
+static void* MergeSummaryOp_Create(TF_OpKernelConstruction* ctx) {}
+
+static void MergeSummaryOp_Delete(void* kernel) {
+  return;
+}
+
+static void MergeSummaryOp_Compute(void* kernel, TF_OpKernelContext* ctx) {
+  tensorflow::Summary s; 
+  std::unordered_set<tensorflow::string> tags; 
+  Status_Wrapper status_wrapper;
+  for (int input_num = 0; input_num < TF_NumInputs(ctx); ++input_num) { 
+    TensorWrapper input_wrapper; 
+    TF_GetInput(ctx, input_num, &input_wrapper.t, status_wrapper.s); 
+    if (TF_GetCode(status_wrapper.s) != TF_OK) {  
+      TF_OpKernelContext_Failure(ctx, status_wrapper.s); 
+      return;
+    }
+
+    auto tags_array = static_cast<tensorflow::tstring*>(
+    		TF_TensorData(input_wrapper.t));
+    for (int i = 0; i < TF_TensorElementCount(input_wrapper.t); ++i) { 
+      const tensorflow::tstring& s_in = tags_array[i]; 
+      tensorflow::Summary summary_in; 
+      if (!tensorflow::ParseProtoUnlimited(&summary_in, s_in)) { 
+        TF_SetStatus(status_wrapper.s, TF_INVALID_ARGUMENT, 
+            "Could not parse one of the summary inputs");
+        TF_OpKernelContext_Failure(ctx, status_wrapper.s);
+        return; 
+      }
+      for (int v = 0; v < summary_in.value_size(); ++v) { 
+        // This tag is unused by the TensorSummary op, so no need to check for 
+        // duplicates.
+        const tensorflow::string& tag = summary_in.value(v).tag(); 
+        if ((!tag.empty()) && !tags.insert(tag).second) { 
+          std::ostringstream err;
+          err << "Duplicate tag " << tag << " found in summary inputs ";
+          TF_SetStatus(status_wrapper.s, TF_INVALID_ARGUMENT, err.str().c_str());
+          TF_OpKernelContext_Failure(ctx, status_wrapper.s);
+          return; 
+        }
+        *s.add_value() = summary_in.value(v); 
+      }
+    }
+  }
+  TensorWrapper summary_tensor_wrapper; 
+  summary_tensor_wrapper.t = TF_AllocateOutput(ctx, 0,
```

Comment:
This is a bit difficult to read; can you use argument comments so that we know what each of the things being passed in is doing? https://google.github.io/styleguide/cppguide.html#Function_Argument_Comments

Reply:
Sorry if the link wasn't clear, here's an example: https://github.com/tensorflow/tensorflow/blob/cf59ede2e4b98b7f4ec868fe9c6c8e6f8dbffee3/tensorflow/cc/experimental/base/tests/tensorhandle_test.cc#L62-L65